src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,412 @@
+
+            // Next, go and update any references to these anonymous type properties to match
+            // the new PascalCased name we've picked for the new properties that will go in
+            // the named type.
+            await ReplacePropertyReferencesAsync(
+                document, editor, containingMember, 
+                propertyMap, cancellationToken).ConfigureAwait(false);
+
+            // Next, go through and replace all matching anonymous types in this method with a call
+            // to construct the new named type we've generated.  
+            await ReplaceMatchingAnonymousTypesAsync(
+                document, editor, className,
+                containingMember, anonymousObject, 
+                anonymousType, cancellationToken).ConfigureAwait(false);
+
+            // Then, actually insert the new class in the appropriate container.
+            var container = anonymousObject.GetAncestor<TNamespaceDeclarationSyntax>() ?? root;
+            editor.ReplaceNode(container, (currentContainer, _) =>
+            {
+                var codeGenService = document.GetLanguageService<ICodeGenerationService>();
+                var options = new CodeGenerationOptions(
+                    generateMembers: true,
+                    sortMembers: false,
+                    autoInsertionLocation: false);
+
+                return codeGenService.AddNamedType(
+                    currentContainer, namedTypeSymbol, options, cancellationToken);
+            });
+
+            var updatedDocument = document.WithSyntaxRoot(editor.GetChangedRoot());
+
+            // Finally, format using the equals+getHashCode service so that our generated methods
+            // follow any special formatting rules specific to them.
+            var equalsAndGetHashCodeService = document.GetLanguageService<IGenerateEqualsAndGetHashCodeService>();
+            return await equalsAndGetHashCodeService.FormatDocumentAsync(
+                updatedDocument, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task ReplacePropertyReferencesAsync(
+            Document document, SyntaxEditor editor, SyntaxNode containingMember,
+            ImmutableDictionary<IPropertySymbol, string> propertyMap, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var identifiers = containingMember.DescendantNodes().OfType<TIdentifierNameSyntax>();
+
+            foreach (var identifier in identifiers)
+            {
+                if (!syntaxFacts.IsNameOfMemberAccessExpression(identifier))
+                {
+                    continue;
+                }
+
+                var symbol = semanticModel.GetSymbolInfo(identifier, cancellationToken).GetAnySymbol() as IPropertySymbol;
```

Comment:
>.GetAnySymbol() [](start = 87, length = 15)

What's a scenario where `GetAnySymbol()` matters here? I would have just grabbed `.Symbol`. If the symbol is unclear, maybe the refactoring shouldn't proceed.

Reply:
so, IDE *generally* (i.e. no hard or fast rule), uses GetAnySymbol to effectively mean: uh... hey compiler, what did you think this was? Oh, it's the symbol i was looking for?  Cool... oh, it's not?  No biggy...

Basically, it's the "what did you think it was, disregarding correct/incorrect?".   In *general*, we let refactoring proceed, with the feeling that if hte compiler thought this was hte symbol we cared about, that's good enough for us.  We don't usually want to block because there might be all sorts of code the compiler isn't certain aobut (i.e. during errors), and having that block helpers like this just ends up confusing to the user due to features not lighting up when users expect them...

So, IMO, this code is fine as is.  It follows the general wishy-washy "just try to do a good job plz" approach we have for most helper tools :)

Reply:
Note: things are different for larger-scale stuff (i.e. rename, or a project/document fix).  Then, not messing up existing code has a higher bar.  But this is not an analyzer.  It's just a local refactoring :)

Reply:
Seems reasonable, but I'm still unclear on scenario where compiler would give such a best guess.

---
In reply to: [198310609](https://github.com/dotnet/roslyn/pull/28066#discussion_r198310609) [](ancestors = 198310609)